### DISCUSSION TITLE
Table rows memoization

### DISCUSSION THREAD

**Discussion by abdullaev388** [rsuite/rsuite#2581] [Q&A] [answered]
Hello. I am using Table component. Each row has 2 controlled inputs. If there are 100 rows the data update is slowing down.

So i would like to have opportunity to memoize rows but with current implementation it is impossible. How can i do that, in fact i want to use React.memo HOC.

**Comment by simonguo** [comment, marked as answer]
Hi @abdullaev388 
You need to check that every input you make causes the Table to re-render, which you should try to avoid. You'd better provide an example of [codesandbox](https://codesandbox.io/s/rsuite-template-5vq6zo2z5l) so we can help you find the right optimization.

The issue will be converted to discussion.

**Reply by abdullaev388** [reply]
Here is the example.
https://codesandbox.io/s/hopeful-swanson-wseq7z?file=/src/index.js

Table with 100 rows, each row has 3 inputs. So please help me how can i optimize it.

**Reply by simonguo** [reply]
Don't update the table's data every time the text is entered.  Customize an InputCell, pass additional incoming Input values.
```tsx
const InputCell = ({ rowData, dataKey, comments, onChange, ...props }) => {
  const value = rowData && comments ? comments[rowData.id - 1][dataKey] : "";
  return (
    <Table.Cell {...props} style={{ padding: "6px" }}>
      <Input
        value={value}
        onChange={(value) => onChange(value, dataKey, rowData.id)}
      />
    </Table.Cell>
  );
};
```

Use InputCell

```tsx
<Table.Column flexGrow={2} align="center" verticalAlign="middle">
    <Table.HeaderCell>Email</Table.HeaderCell>
    <InputCell
      dataKey="email"
      comments={comments}
      onChange={onCommentChange}
    />
</Table.Column>
```

Complete example: https://codesandbox.io/s/inspiring-bohr-6wvlbw

**Reply by abdullaev388** [reply]
Thank you sir, you really made it work better.
But one more question please. Even though i change only 1 input all others are also being re-rendered.
If you press F12 and enable throttling, it will still have kind of bad performance.
What can you recommend to do in this case?

**Reply by simonguo** [reply]
There are several suggestions:
- Rendered as an Input only when dealing with editing state, plain text is displayed by default. https://rsuitejs.com/components/table/#editable
- It is best not to have too much table data, you can use paging display. https://rsuitejs.com/components/table/#pagination
- Do not maintain state, make Input uncontrolled, thus avoiding Table re-rendering. (Not recommended)
